pkg/vmconfig/crypto/crypto_reconciler.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package crypto
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha3"
+	"github.com/vmware-tanzu/vm-operator/pkg/bitmask"
+	"github.com/vmware-tanzu/vm-operator/pkg/conditions"
+	ctxgen "github.com/vmware-tanzu/vm-operator/pkg/context/generic"
+	"github.com/vmware-tanzu/vm-operator/pkg/vmconfig"
+	"github.com/vmware-tanzu/vm-operator/pkg/vmconfig/crypto/internal"
+)
+
+type reconciler struct{}
+
+var _ vmconfig.Reconciler = reconciler{}
+
+// New returns a new Reconciler for a VM's crypto state.
+func New() vmconfig.ReconcilerWithContext {
+	return reconciler{}
+}
+
+// Name returns the unique name used to identify the reconciler.
+func (r reconciler) Name() string {
+	return "crypto"
+}
+
+func (r reconciler) WithContext(parent context.Context) context.Context {
+	return ctxgen.WithContext(
+		parent,
+		internal.ContextKeyValue,
+		func() internal.State { return internal.State{} })
+}
+
+// SprintfStateNotSynced formats and returns the message for when the encryption
+// state cannot be synced.
+func SprintfStateNotSynced(op string, msgs ...string) string {
+	if len(msgs) == 0 {
+		return ""
+	}
+	var msg string
+	switch len(msgs) {
+	case 1:
+		msg = msgs[0]
+	case 2:
+		msg = fmt.Sprintf("%s and %s", msgs[0], msgs[1])
+	default:
+		msg = fmt.Sprintf(
+			"%s, and %s",
+			strings.Join(msgs[:len(msgs)-1], ", "),
+			msgs[len(msgs)-1])
+	}
+	return fmt.Sprintf("Must %s when %s vm", msg, op)
+}
+
+// Reason is the type used by reasons given to the condition
+// vmopv1.VirtualMachineEncryptionSynced.
+type Reason uint8
+
+const (
+	ReasonEncryptionClassNotFound Reason = 1 << iota
+	ReasonEncryptionClassInvalid
+	ReasonInvalidState
+	ReasonInvalidChanges
+	ReasonReconfigureError
+	_reasonMax = 1 << (iota - 1)
+)
+
+func (r Reason) MaxValue() Reason {
+	return _reasonMax
+}
+
+func (r Reason) StringValue() string {
+	switch r {
+	case ReasonEncryptionClassNotFound:
+		return "EncryptionClassNotFound"
+	case ReasonEncryptionClassInvalid:
+		return "EncryptionClassInvalid"
+	case ReasonInvalidState:
+		return "InvalidState"
+	case ReasonInvalidChanges:
+		return "InvalidChanges"
+	case ReasonReconfigureError:
+		return "ReconfigureError"
+	}
+	return ""
+}
+
+func (r Reason) String() string {
+	return bitmask.String(r)
```

Comment:
Andrew, I would rather have us just use the reasons explicitly, as opposed to define a type for them, and then use bitmasks to figure out the reasons that apply.  This will be hard to maintain and I don't see many usecases for this pattern. 

My recommendation would be to shelve the bitmask manipulation entirely and references the reasons explicitly where needed.

Reply:
The reasons *are* used explicitly, but there can be many different kinds and combinations. So I disagree there is not a use case for this in the future. There is literally a use case for it in this PR. The documentation in `vm.md` clearly outlines the reasons to the user. I appreciate the feedback, but I am going to keep this as-is.

Reply:
FWIW, here is [a test case](https://github.com/vmware-tanzu/vm-operator/pull/673/files#diff-916fea486b6a687a518cffddb850c48ae19a10f9f42864ac3e3507126a6cc233R324-R341) that shows when the reason can be both `InvalidState` _and_ `InvalidChanges`

Reply:
All that said, this _is_ something I am willing to revisit. I will give it more thought, but I am not going to let it block this PR. After it merges, I'll sit down and consider how many possible combinations there can be in practical terms. If it is just `InvalidState` and `InvalidChanges`, then I can have a distinct reason for that. Let me look.